### ISSUE TITLE
How does tokenizers tokenize non-english characters into tokens?

### ISSUE THREAD

**Issue by dongrixinyu** [huggingface/tokenizers#1162]
Here is some examples I tried using tokenizers:

```
from transformers import GPT2TokenizerFast
tokenizer = GPT2TokenizerFast.from_pretrained("gpt2")
tokenizer("是")
# {'input_ids': [42468], 'attention_mask': [1]}
tokenizer("我说你倒是快点啊")
# {'input_ids': [22755, 239, 46237, 112, 19526, 254, 161, 222, 240, 42468, 33232, 104, 163, 224, 117, 161, 243, 232], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]}
tokenizer("東")
# {'input_ids': [30266, 109], 'attention_mask': [1, 1]}
tokenizer("東京")
# {'input_ids': [30266, 109, 12859, 105], 'attention_mask': [1, 1, 1, 1]}
tokenizer("東京メトロ")
# {'input_ids': [30266, 109, 12859, 105, 26998, 13298, 16253], 'attention_mask': [1, 1, 1, 1, 1, 1, 1]}
tokenizer("メトロ")
# {'input_ids': [26998, 13298, 16253], 'attention_mask': [1, 1, 1]}
tokenizer("This is my fault")
{'input_ids': [1212, 318, 616, 8046], 'attention_mask': [1, 1, 1, 1]}
```

The code above is some examples I tried.
The last example is an english sentence and I can understand that `This` corresponds to `"This":1212` in the `vocab.json`, ` is` corresponds to `"\u0120is": 318`.

But I can not understand why this tool tokenizes non-english sequence into some tokens I can not find in the vocab.
For example: 
`東` is been tokenized into `30266 and 109`. The results in the `vocab.json` is `"æĿ":30266` and `"±":109`.
`メ` is been tokenized into `26998`. The results in the `vocab.json` is `"ãĥ¡":26998`.

I searched the huggingface documents and website and find no clue.

So could you help me figure out why? 
Thank you very much.

**Comment by roudimit** [comment]
I was also wondering about this. It turns out that GPT-2 uses a Byte-level BPE tokenizer, so it can encode any string. See these posts for more details:
https://github.com/huggingface/tokenizers/issues/203#issuecomment-605105611
https://stackoverflow.com/questions/69089843/why-bpe-encoding-trained-on-english-and-applied-on-bengali-doesnot-return-unknow
https://github.com/huggingface/tokenizers/issues/889#issuecomment-1019919837
https://medium.com/@pierre_guillou/byte-level-bpe-an-universal-tokenizer-but-aff932332ffe

I have a question about the particular token IDs selected through the encoding process. Is is possible to see how they relate to the bytes of the string being encoded?

For example,
```
>>> tokenizer.encode('東')
[30266, 109]
>>> '東'.encode('utf-8')
b'\xe6\x9d\xb1'
>>> tokenizer.vocab["æĿ"]
30266
>>> tokenizer.vocab["±"]
109
```
I guess "æĿ" and "±" are selected since the token IDs correspond to the bytes in ('東').
But is it possible to see how they correspond to b'\xe6\x9d\xb1'? It's a little confusing because "æĿ" and "±" have six bytes in total, while '東' has three bytes.
```
>>> "æĿ".encode('utf-8')
b'\xc3\xa6\xc4\xbf'

>>> "±".encode('utf-8')
b'\xc2\xb1'
```

**Comment by dongrixinyu** [comment]
Thank you very much. 

That's confusing me too. The bytes encoded from `東` and `æĿ` have no common parts. At first I have the same guess of byte-level tokenization. But the three byte char `東` and six byte string `æĿ` makes me assume that the `tokenizers` treats these characters as `UNK`s. Cause for GPT2, most of the texts is in English. Bangali, Chinese and Japanese just compose only a small fraction of the total corpus.

According to your reply, I could infer that the sentence with or without spaces could generate different sequence of tokens in ChatGPT like below:
<img width="538" alt="f46a1ced45ffaa3373e80ecf1bbd2a4" src="https://user-images.githubusercontent.com/27001062/220043638-c69d1d9f-191c-4c65-a12c-4b47b022a58f.png">

But it can generate similar replies. That's a little bit out of my mind.

**Comment by roudimit** [comment]
Hi @Narsil thanks for the reply.
I decided to check the original GPT-2 code (https://github.com/openai/gpt-2/blob/a74da5d99abaaba920de8131d64da2862a8f213b/src/encoder.py), which was helpful to understand more of the details.
I made this Google Colab with my findings: https://colab.research.google.com/drive/15tMASZ0NLm8bnxkM4uXCRgdzznSpbp9L?usp=sharing
@ dongrixinyu In the Colab, I show more details about your example ('東').

**Comment by dongrixinyu** [comment]
@roudimit @Narsil Really thank both of you guys. Your example is really clear and expressive. 

So I know `�` is actually an unknown printable token which is rarely occurred in the corpus. I will add this Byte-level BPE into my python lib `jionlp`

**Comment by Narsil** [comment]
> BTW, I am curious of the disadvantages of Byte-level BPE when applyed in GPTs.

None really. It's not as optimal for some compared to Unigram. Also Unigram can have longer range (BPE is limited to whitespace splitting so hard to work directly on Chinese, you need some pretokenization to have reasonable training times). 
Unigram can have dropout.

However the advantages of Unigram compared to its complexity have not warranted everyone to switch over (yet). My guess is we'll have tokenizerless models before Unigram takes over.

**Comment by dongrixinyu** [comment]
I have read the `tokenizers` course on the huggingface thoroughly today.

I think the tokenizer which can smoothly processing multi-languages is the most significant feature. So `SentencePiece`  is combined with all kinds of subwords algorithms.

Byte-level BPE cuts Chinese, Japanese and all other non-English characters into something with no clear semantic meaning. I think this is not that bad. Because like the example I showed above concerning ChatGPT. The model can totally understand what I mean, although the tokenization results is different. According to this, I assume Unigram providing a more precise tokenization results is not really very important only if the params of the model is large enough.

**Comment by Alex-Songs** [comment]
@dongrixinyu 哈喽大佬，我这边用bytelevel bpe解码中文时出现�："没 有 � 加 价 的 行 为"。这是因为我训练bytelevel bpe model时用的语料不够多，造成有些词解码不出来嘛？

**Comment by Narsil** [comment]
My mandarin is not that great, byte level is using bytes, so if you have � symbol, which is the unknown utf-8 it simply means the model generated some bytes which were not valid utf-8. This is perfectly normal from `tokenizers` standpoint.

For you model, you may need more training indeed, or this data is a bit out of domain. More rarely, you could have an issue in your dataset itself (for instance if it's not encoded in utf-8 and some stuff gets lost in translation).

Please try to stick to english here (even if automated translation, keep at least an english version for readers.)

**Comment by Alex-Songs** [comment]
@Narsil Thanks
